sql/api/src/main/scala/org/apache/spark/sql/types/StructField.scala

```diff
@@ -63,7 +66,60 @@ case class StructField(
     ("name" -> name) ~
       ("type" -> dataType.jsonValue) ~
       ("nullable" -> nullable) ~
-      ("metadata" -> metadata.jsonValue)
+      ("metadata" -> metadataJson)
+  }
+
+  private def metadataJson: JValue = {
+    val metadataJsonValue = metadata.jsonValue
+    metadataJsonValue match {
+      case JObject(fields) if collationMetadata.nonEmpty =>
+        val collationFields = collationMetadata.map(kv => kv._1 -> JString(kv._2)).toList
+        JObject(fields :+ (DataType.COLLATIONS_METADATA_KEY -> JObject(collationFields)))
+
+      case _ => metadataJsonValue
+    }
+  }
+
+  /** Map of field path to collation name. */
+  private lazy val collationMetadata: mutable.Map[String, String] = {
+    val fieldToCollationMap = mutable.Map[String, String]()
+
+    def visitRecursively(dt: DataType, path: String): Unit = dt match {
+      case at: ArrayType =>
+        processDataType(at.elementType, path + ".element")
+
+      case mt: MapType =>
+        processDataType(mt.keyType, path + ".key")
+        processDataType(mt.valueType, path + ".value")
+
+      case st: StringType if isCollatedString(st) =>
+        fieldToCollationMap(path) = collationName(st)
+
+      case _ =>
+    }
+
+    def processDataType(dt: DataType, path: String): Unit = {
+      if (isCollatedString(dt)) {
+        fieldToCollationMap(path) = collationName(dt)
+      } else {
+        visitRecursively(dt, path)
+      }
+    }
+
+    visitRecursively(dataType, name)
```

Comment:
This is per field metadata, do we really need to put the filed name in the path key? My worry is that the field name can contain special chars, e.g. `name.key`, which may cause issues without proper quoting. How about we use empty string as the path key for the top-level field?

Reply:
I added a test to be sure this works, but I don't see why we would need any special logic for this case.

Even if we have a column `name.key` this will never be ambiguous in the metadata as the struct field could be a string or a map but never both, so parsing will just work regardless.

Reply:
I added a test to be sure this works, but I don't see why we would need any special logic for this case.

Even if we have a column name.key this will never be ambiguous in the metadata as the struct field could be a string or a map but never both, so parsing will just work regardless - we could never have a case of adding the collation two times.

Reply:
what if the field is named `key.element.value`?

Reply:
Each column has separate metadata so map and string can never clash -> here is how it would look like in json:

```json
{
  "type": "struct",
  "fields": [
    {
      "name": "c1.key",
      "type": "string",
      "nullable": true,
      "metadata": {
        "__COLLATIONS": {
          "c1.key": "ICU.UNICODE_CI"
        }
      }
    },
    {
      "name": "c1",
      "type": {
        "type": "map",
        "keyType": "string",
        "valueType": "string",
        "valueContainsNull": true
      },
      "nullable": true,
      "metadata": {
        "__COLLATIONS": {
          "c1.value": "ICU.UNICODE",
          "c1.key": "ICU.UNICODE"
        }
      }
    }
  ]
}
```

"c1.key" will mean different things in each column - for the first it's just the entire column and for the second it is the key of the map